[dec 12 merge] fix: add cross links to the packaging guide (#431)
pyOpenSci/python-package-guide: package-structure-code/python-package-structure.md

Note the location of the following directories in the example above:

- **docs/:** discussed in our docs chapter, this directory contains your user-facing documentation website. In a **src/** layout docs/ are normally included at the same directory level as the **src/** folder.

> - **tests/** This directory contains the tests for your project code. In a **src/** layout, tests are normally included at the same directory level as the **src/** folder.

Reviewer: You capitalized this bullet, but no others. Not sure what your preference is.

Author: this happens because i never know when to capitalize and when to not capitalize! let's capitalize?

## The src/ layout and testing

The benefit of using the **src/package** layout is that it ensures tests are run against the
installed version of your package rather than the files in your package
working directory. If you run your tests on your files rather than the
installed version of your package, you may be missing issues that users encounter when
your package is installed.

> If `tests/` are outside the **src/package** directory, they aren't included in the package [wheel](python-wheel). This makes your package size slightly smaller, which places a smaller storage burden on PyPI, which has over 400,000 packages to support.

Reviewer: ```suggestion
If `tests/` are outside the **src/package** directory, they aren't included in the package's [wheel](python-wheel). This makes your package size slightly smaller, which places a smaller storage burden on PyPI, and makes them faster to fetch.
```

This means that if you run your tests in your package's working directory, using a flat layout, `/package/module.py`, Python will discover `package/module.py` file before it discovers the installed package.

However, if your package lives in a src/ directory structure **src/package**, then it won't be added to the Python path by default. This means that when you import your package, Python will be forced to search the active environment (which has your package installed).

Note: Python versions 3.11 and above have a path setting that can be adjusted to ensure the priority is to use installed packages first (e.g., `PYTHONSAFEPATH`).
```

### Don't include tests in your package wheel

> Writing [tests](/tests/index.html) for your package is important; however, we do not recommend including tests as part of your [package wheel](python-wheel) by default. However, not including tests in your package distribution will make it harder for people other than yourself to test whether your package runs properly on their system. If you have a small test suite (Python files + data), and think your users may want to run tests locally on their systems, you can include tests by moving the `tests/` directory into the **src/package** directory (see example below).

Author: thank you for catching all of these issues!! CI is now happy thanks to you!!

docs/
```

Including the **tests/** directory in your **src/package** directory ensures that tests will be included in your package's [wheel](python-wheel).

Be sure to read the [pytest documentation for more about including tests in your package distribution](https://docs.pytest.org/en/7.2.x/explanation/goodpractices.html#choosing-a-test-layout-import-rules).

```{admonition} Challenges with including tests and data in a package wheel
:class: tip

Tests, especially when accompanied by test data, can create a few small challenges, including:

- Take up space in your distribution, which will build up over time as storage space on PyPI
- Large file sizes can also slow down package installation.

However, in some cases, particularly in the scientific Python ecosystem, you may need to include tests.
```

### **Don't include test suite datasets in your package**

If you include your tests in your package distribution, we strongly
discourage you from including data in your test suite directory. Rather,
host your test data in a repository such as Figshare or Zenodo. Use a
tool such as [Pooch](https://www.fatiando.org/pooch/latest/) to access
the data when you (or a user) runs tests.

> For more information about Python package tests, see the [tests section of our guide](/tests/index.html).

Author: Also fixed 🚀

- The **src/package** layout is semantically more clear. Code is always found in the
  **src/package** directory, `tests/` and `docs/`are in the root directory.

```{important}
If your package tests require data, do NOT include that
data within your package structure. Including data in your package structure increases the size of your
distribution files. This places a maintenance toll on repositories like PyPI and

> Anaconda.org has to deal with thousands of package uploads.

Author: ```suggestion
Anaconda.org that have to deal with thousands of package uploads.
```

> Anaconda.org has to deal with thousands of package uploads.

Author: Fixed!!

:color: success
:class: sd-rounded-pill float-left

Click here for a quickstart tutorial on creating your Python package.

:::

(flat-layout)=
## What is the flat Python package layout?

Many scientific packages use the **flat-layout** given:

- This layout is used by many core scientific Python packages such as NumPy, SciPy, and Matplotlib.
- Many Python tools depend upon tools in other languages and/or complex builds
  with compilation steps. Many maintainers prefer features
  of the flat layout for more complex builds.

While we suggest that you use the src/ layout discussed above, it's important to also
understand the flat layout, especially if you plan to contribute to a package that uses this layout

```{admonition} Why most scientific Python packages do not use source
:class: tip

In most cases, moving to the **src/package** layout for
larger scientific packages that already use a flat layout would consume significant time.

However, the advantages of using the  **src/package** layout for a beginner are significant.
As such, we recommend that you use the **src/package** layout if you are creating a new package.

Numerous packages in the ecosystem [have had to move to a

> src/ layout](https://github.com/scikit-build/cmake-python-distributions/pull/145)

Reviewer 2: ```suggestion
**src/package** layout](https://github.com/scikit-build/cmake-python-distributions/pull/145).
```

```{admonition} Multiple packages in a src/ folder
:class: tip

> In some more advanced cases, you may have more than one package in your src/ directory. See [black's GitHub repo](https://github.com/psf/black/tree/main/src) for an example of this. However, for most beginners you will likely only have one sub-directory in your **src/** folder.

Reviewer 2: ```suggestion
In some more advanced cases, you may have more than one package in your **src/** directory. See [Black's GitHub repo](https://github.com/psf/black/tree/main/src) for an example of this. However, for most beginners you will likely only have one sub-directory in your **src/** folder.
```

``` -->

<!--
```{admonition} A few notes about the src/ layout
:class: tip

> It is important to note here that sometimes, when using the src/package structure, the directory name (e.g., package name) is different from the actual project or package name. What is important to take away here is that you should store your code within a subdirectory within **src/**.

Reviewer 2: ```suggestion
It is important to note here that sometimes, when using the **src/package** structure, the directory name (e.g., package name) is different from the actual project or package name. What is important to take away here is that you should store your code within a subdirectory within **src/**.
```
